homeassistant/components/hlk_sw16.py

```diff
@@ -0,0 +1,461 @@
+"""
+Support for HLK-SW16 relay switch.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/hlk-sw16/
+"""
+import asyncio
+from collections import deque
+import logging
+import binascii
+import codecs
+import async_timeout
+
+import voluptuous as vol
+
+from homeassistant.const import (
+    CONF_HOST, CONF_PORT,
+    EVENT_HOMEASSISTANT_STOP, CONF_SWITCHES, CONF_NAME)
+from homeassistant.core import CoreState, callback
+from homeassistant.exceptions import HomeAssistantError
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.discovery import load_platform
+from homeassistant.helpers.dispatcher import (
+    async_dispatcher_send, async_dispatcher_connect)
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_EVENT = 'event'
+ATTR_STATE = 'state'
+
+CONF_RECONNECT_INTERVAL = 'reconnect_interval'
+
+DATA_DEVICE_REGISTER = 'hlk_sw16_device_register'
+DATA_ENTITY_LOOKUP = 'hlk_sw16_entity_lookup'
+DEFAULT_RECONNECT_INTERVAL = 10
+CONNECTION_TIMEOUT = 10
+
+DOMAIN = 'hlk_sw16'
+
+SIGNAL_AVAILABILITY = 'hlk_sw16_device_available'
+SIGNAL_HANDLE_EVENT = 'hlk_sw16_handle_event_{}'
+
+TMP_ENTITY = 'tmp.{}'
+
+SWITCH_SCHEMA = vol.Schema({
+    vol.Optional(CONF_NAME): cv.string,
+})
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_PORT): vol.Any(cv.port, cv.string),
+        vol.Optional(CONF_HOST): cv.string,
+        vol.Optional(CONF_RECONNECT_INTERVAL,
+                     default=DEFAULT_RECONNECT_INTERVAL): int,
+        vol.Required(CONF_SWITCHES): vol.Schema({cv.slug: SWITCH_SCHEMA}),
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+
+class SW16Protocol(asyncio.Protocol):
```

Comment:
Device specific I/O and data parsing should be moved to standalone library published on pypi.

https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#6-communication-with-devices-services

Reply:
Yeah, I wasn't sure if it made sense to have generalized library for this protocol since it is fairly simple and my implementation is fairly tightly coupled to home-assistant and doesn't implement a number of device features that would be difficult to map to home-assistant(such as the internal timer/clock stuff). Should I be creating a separate library even if it's designed to be only used by home-assistant?

Reply:
Yes. But preferably make it as general as possible. And don't make it aware of home assistant.